libcontainer/rootfs_linux.go

```diff
@@ -741,7 +741,19 @@ func prepareRoot(config *configs.Config) error {
 }
 
 func setReadonly() error {
-	return unix.Mount("/", "/", "bind", unix.MS_BIND|unix.MS_REMOUNT|unix.MS_RDONLY|unix.MS_REC, "")
+	flags := uintptr(unix.MS_BIND | unix.MS_REMOUNT | unix.MS_RDONLY | unix.MS_REC)
+
+	err := unix.Mount("", "/", "", flags, "")
```

Comment:
I don't understand why this mount is needed (or, in other words, why it has to be done using two mount(2) calls).

Reply:
The logic is that we only retain the old mount flags if it fails, but I'm confused why this code belongs in runc. I solved this problem in Docker some time ago with this code:

```go
// Get the set of mount flags that are set on the mount that contains the given
// path and are locked by CL_UNPRIVILEGED. This is necessary to ensure that
// bind-mounting "with options" will not fail with user namespaces, due to
// kernel restrictions that require user namespace mounts to preserve
// CL_UNPRIVILEGED locked flags.
func getUnprivilegedMountFlags(path string) ([]string, error) {
	var statfs unix.Statfs_t
	if err := unix.Statfs(path, &statfs); err != nil {
		return nil, err
	}

	// The set of keys come from https://github.com/torvalds/linux/blob/v4.13/fs/namespace.c#L1034-L1048.
	unprivilegedFlags := map[uint64]string{
		unix.MS_RDONLY:     "ro",
		unix.MS_NODEV:      "nodev",
		unix.MS_NOEXEC:     "noexec",
		unix.MS_NOSUID:     "nosuid",
		unix.MS_NOATIME:    "noatime",
		unix.MS_RELATIME:   "relatime",
		unix.MS_NODIRATIME: "nodiratime",
	}

	var flags []string
	for mask, flag := range unprivilegedFlags {
		if uint64(statfs.Flags)&mask == mask {
			flags = append(flags, flag)
		}
	}

	return flags, nil
}
```

The problem with doing it in runc is that we silently add mount flags which the user didn't specify. In theory this won't reduce security (since all of the `MNT_LOCKED` flags restrict the mount rather than permitting extra operations) but I still don't like the look of this.

@giuseppe did you hit this issue with podman? Maybe it'd be better to port the code from Docker to podman?

Reply:
As an aside, `MS_BIND | MS_REMOUNT | MS_RDONLY | MS_REC` doesn't actually make the mount recursively read-only but that's a separate issue...

Reply:
@cyphar we are currently hitting it with CRI-O here: https://github.com/cri-o/cri-o/pull/4333

I am fine to move the check into CRI-O but it is currently affecting the rootfs remount, so we'd need a way to pass down these options in the OCI configuration.  AFAICS, we can set only the rootfs propagation

Reply:
Indeed, I don't see a way to specify mount options for container root mount.

Reply:
Yeah if this is for root mounts this is a bit more of an issue -- but since this code only affects `/` remounting this might actually be alright since that means a user didn't specify it...

Reply:
Should we drop `MS_REC` while at it? It does not make sense anyway

Reply:
> Should we drop MS_REC while at it? It does not make sense anyway

can be addressed separately